### PR TITLE
feat: (authentication) Api key based authentication [#2000]

### LINKED ISSUE #2000: [feature]: Api Key based authentication
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Summary

Hello,

We see many apis and api gateways using api key as a standard authentication mechanism, it would be awesome to support this authentication method in hoppscotch. This would require the ability to specificy the place where to set the api key (header or path) and under which name (different vendor and implementation use different header name).

### Why should this be worked on?

Currently APIs using API Keys passed through headers have their authentication specified as any other header when they actually concern Authentication. To have a consistent handling of Authentication it should be made possible to specify API Key as an autentication method in hoppscotch

### PR DESCRIPTION
**Pull request by EdwinFajardoBarrera**
Closes #2000  

### Description

This is the PR linked to the addition of support api-key based authentication.

### Checks

- [x] My pull request adheres to the code style of this project
- [ ] My code requires changes to the documentation
- [ ] I have updated the documentation as required
- [x] All the tests have passed

### Additional Information

Added:
image
image

image

### CODE REVIEW

**Comment by liyasthomas** [comment]
@EdwinFajardoBarrera I've shared my suggestions on this PR, do let me know if those sound good to you. Except transferring the logic of appending the parameter to the actual parameters list at the time of sending the request all looks good to me. Because if we append the parameter to the actual parameters section and choose to change the `Pass by` value again, the added parameter still stays in the actual parameters section. The ideal solution is to not add parameters to the actual parameters section instead add them to effective parameters. More info in the review description. Thank you.

**Comment by EdwinFajardoBarrera** [comment]
I think now it works well, using `effectiveFinalParams` avoided all the problems. I would like to read your comments in order to create the PR.

**Review by liyasthomas — CHANGES_REQUESTED** [review]

**Review by liyasthomas — APPROVED** [review]
LGTM

File: packages/hoppscotch-app/helpers/utils/EffectiveURL.ts [hunk 1]
```diff
@@ -100,6 +112,22 @@ export function getEffectiveRESTRequest(
...
         )}`,
       })
+    } else if (request.auth.authType === "api-key") {
+      const { key, value, addTo } = request.auth
+
+      if (addTo === "Header") {
+        effectiveFinalHeaders.push({
+          active: true,
+          key: parseTemplateString(key, envVariables),,
+          value: parseTemplateString(value, envVariables),
```
Comment (liyasthomas) [inline packages/hoppscotch-app/helpers/utils/EffectiveURL.ts:122]:
```suggestion
          key: parseTemplateString(key, envVariables),
          value: parseTemplateString(value, envVariables),
```

File: packages/hoppscotch-app/helpers/utils/EffectiveURL.ts [hunk 2]
```diff
@@ -100,6 +112,22 @@ export function getEffectiveRESTRequest(
...
+        })
+      } else if (addTo === "Query params") {
+        effectiveFinalParams.push({
+          active: true,
+          key,
+          value,
```
Comment (liyasthomas) [inline packages/hoppscotch-app/helpers/utils/EffectiveURL.ts:128]:
```suggestion
          key: parseTemplateString(key, envVariables),
          value: parseTemplateString(value, envVariables),
```

**Review by liyasthomas — CHANGES_REQUESTED** [review]

**Review by liyasthomas — APPROVED** [review]

**Pull request merged**